code/modules/antagonists/mind_flayer/flayer_datum.dm

```diff
@@ -0,0 +1,295 @@
+	var/list/category_stage = list(CATEGORY_GENERAL = 1, CATEGORY_DESTROYER = 1, CATEGORY_INTRUDER = 1)
+	/// If the mindflayer can still pick a stage 4 ability
+	var/can_pick_capstone = TRUE
+
+/proc/ismindflayer(mob/M)
+	return M.mind?.has_antag_datum(/datum/antagonist/mindflayer)
+
+/datum/antagonist/mindflayer/New()
+	. = ..()
+	if(!length(ability_list))
+		ability_list = get_spells_of_type(FLAYER_PURCHASABLE_POWER)
+		ability_list += get_passives_of_type(FLAYER_PURCHASABLE_POWER)
+
+/datum/antagonist/mindflayer/Destroy(force, ...)
+	QDEL_NULL(owner.current?.hud_used?.vampire_blood_display)
+	remove_all_abilities()
+	remove_all_passives()
+	..()
+
+// This proc adds extra things, and base abilities that the mindflayer should get upon becoming a mindflayer
+/datum/antagonist/mindflayer/on_gain()
+	. = ..()
+	var/list/innate_powers = get_spells_of_type(FLAYER_INNATE_POWER)
+	for(var/power_path as anything in innate_powers)
+		var/datum/spell/flayer/to_add = new power_path
+		add_ability(to_add, src)
+	owner.current.faction += list("flayer") // In case our robot is mindlessly spawned in somehow, and they won't accidentally kill us
+
+/datum/antagonist/mindflayer/give_objectives()
+	add_antag_objective(/datum/objective/swarms)
+	forge_basic_objectives()
+
+/datum/antagonist/mindflayer/proc/get_swarms()
+	return usable_swarms
+
+/datum/antagonist/mindflayer/proc/set_swarms(amount, update_total = FALSE) //If you adminbus someone's swarms it won't add or remove to the total
+	var/old_swarm_amount = usable_swarms
+	usable_swarms = amount
+	if(update_total)
+		var/difference = usable_swarms - old_swarm_amount
+		total_swarms_gathered += difference
+
+/*
+* amount - The positive or negative number to adjust the swarm count by, result clamped above 0
+*/
+/datum/antagonist/mindflayer/proc/adjust_swarms(amount, bound_lower = 0, bound_upper = INFINITY)
+	set_swarms(directional_bounded_sum(usable_swarms, amount, bound_lower, bound_upper), TRUE)
+
+//This is mostly for flavor, for framing messages as coming from the swarm itself. The other reason is so I can type "span" less.
+/datum/antagonist/mindflayer/proc/send_swarm_message(message)
+	if(HAS_TRAIT(owner.current, TRAIT_MINDFLAYER_NULLIFIED))
+		message = stutter(message, 0, TRUE)
+	to_chat(owner.current, "<span class='sinister'>[message]</span>")
+
+/**
+	Checks for any reason that you should not be able to drain someone for.
+	Returns either true or false, if the harvest will work or not.
+*/
+/datum/antagonist/mindflayer/proc/check_valid_harvest(mob/living/carbon/human/H)
+	if(HAS_TRAIT(owner.current, TRAIT_MINDFLAYER_NULLIFIED))
+		send_swarm_message("We do not have the energy for this...")
+		return FALSE
+
+	var/obj/item/organ/internal/brain/brain = H.get_int_organ(/obj/item/organ/internal/brain)
+	if(!istype(brain))
+		send_swarm_message("This entity has no brain to harvest from.")
+		return FALSE
+	if(!H.ckey && !H.player_ghosted)
```

Comment:
I think you probably want to check mind here, rather than ckey. Otherwise someone could combat log you and prevent you from harvesting them.